File: 3학년 겨울방학/Good Bye, BOJ 2021!/성싶당 밀키트.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef long long Long;

int main() {
    std::cin.tie(nullptr)->sync_with_stdio(false);

    Long n, g, k;
    std::cin >> n >> g >> k;

    std::vector<std::pair<Long, Long>> persistent, removable;
    for (int i = 0; i < n; ++i) {
        Long s, l, o;
        std::cin >> s >> l >> o;
        (o == 1LL ? removable : persistent).emplace_back(s, l);
    }

    auto isPossible = [&](Long t) -> bool {
        __int128_t sum = 0;
        for (auto p : persistent) {
            sum += __int128_t(p.first) * std::max(Long(1), t-p.second);
            if (sum > g) {
                return false;
            }
        }

        std::vector<__int128_t> cands;
        for (auto p : removable) {
            cands.push_back(__int128_t(p.first) * std::max(Long(1), t-p.second));
        }
        std::sort(cands.begin(), cands.end());
        for (int i = 0; i < Long(cands.size())-k && k < cands.size(); ++i) {
            sum += cands[i];
            if (sum > g) {
                return false;
            }
        }
        
        return true;
    };

    Long left = 0, right = INT64_MAX;
    while (left+1 < right) {
        Long mid = (left+right)/2;
        if (isPossible(mid)) {
            left = mid;
        } else {
            right = mid;
        }
    }

    if (isPossible(right)) {
        std::cout << right;
    } else {
        std::cout << left;
    }

    return 0;
}

File: 2019-2021/3학년 겨울방학/Good Bye, BOJ 2021!/성싶당 밀키트.py
import sys
input = sys.stdin.readline

n, g, k = list(map(int, input().split(" ")))

persistent, removable = [], []
for i in range(n):
    s, l, o = list(map(int, input().split(" ")))
    if o == 1:
        removable.append([s, l])
    else:
        persistent.append([s, l])

def is_possible(t):
    sum = 0
    for s, l in persistent:
        sum += s * max(1, t-l)
        if sum > g:
            return False
    
    cands = []
    for s, l in removable:
        cands.append(s * max(1, t-l))
    cands.sort()
    if k < len(cands):
        for i in range(len(cands)-k):
            sum += cands[i]
            if sum > g:
                return False
    
    return True


left = 0; right = int(1e18)
while left+1 < right:
    mid = (left+right)//2
    if is_possible(mid):
        left = mid
    else:
        right = mid

if is_possible(right):
    print(right)
else:
    print(left)